design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
+
+We may want to introduce a limit to the number of rows that may be returned from these `executeQuery()` methods,
+in order to decrease the likelihood of running out of memory.
+
+### Insert values
+
+A method to insert one row at a time:
+```
+  /**
+   * Insert a single row into the relevant stream/table.
+   *
+   * @param streamName name of stream/table.
+   * @param row the row to insert.
+   * @return a future that completes once the request has been processed.
+   */
+  CompletableFuture<Void> insertInto(String streamName, Map<String, Object> row);
+```
+
+A method to stream inserts (via the `/inserts-stream` endpoint):
+```
+  CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<List<Object>> insertsPublisher);
+}
+```
+where `InsertResponse` is as follows:
+```
+public interface InsertResponse {
+
+  /**
+   * Whether the row was successfully inserted or not.
+   */
+  boolean isSuccessful();
+
+  /**
+   * Unique sequence number for the row in the stream of inserts.
+   */
+  int getSequenceNum();
+
+  /**
+   * If unsuccessful, the error message.
+   */
+  String getErrorMessage();
+
+  /**
+   * If unsuccessful, the error code.
+   */
+  int getErrorCode();
+
+}
+```
+
+### DDL/DML operations
+
+#### `CREATE <STREAM/TABLE>`, `CREATE <STREAM/TABLE> ... AS SELECT`, `INSERT INTO`
+
+`Client` methods:
+```
+  /**
+   * Execute DDL statement: `CREATE STREAM`, `CREATE TABLE`
+   */
+  CompletableFuture<DdlResponse> executeDdl(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql, Map<String, Object> properties);
+
+```
+with
+```
+public interface DdlResponse {
+
+  enum Status { QUEUED, PARSING, EXECUTING, RUNNING, TERMINATED, SUCCESS, ERROR }
+  
+  Status getStatus();
+  
+  String getStatusMessage();
+  
+  long getCommandSequenceNumber();
```

Comment:
Why do we need command sequence number, command ID, etc? Are they used elsewhere in the API?

Reply:
```
{ QUEUED, PARSING, EXECUTING, RUNNING, TERMINATED, SUCCESS, ERROR }
```
Why does the user care about these states? I think I'd only be interested in whether the command succeeded or not, and that can be signalled with a CompletableFuture, with the exception containing the error if it didn't succeeed.

Reply:
As an overall principle I like to think that "less is more" when it comes to API design. I.e. don't add stuff unless it's super important - as too much information can be overwhelming to users.

Reply:
Good catch. I think it's worth exposing the command sequence number option for requests (in addition to `sql` which is required and properties which is optional), as that's the mechanism used by ksqlDB to ensure earlier requests have been executed before later requests.

Command ID is used by the `/status` endpoint, which isn't planned to be supported by the Java client as part of this KLIP. I think it's unlikely users will want to use the `/status` endpoint on their own (outside of the client) so I'll remove command ID.

Reply:
> Why does the user care about these states? I think I'd only be interested in whether the command succeeded or not, and that can be signalled with a CompletableFuture, with the exception containing the error if it didn't succeeed.

A status of `SUCCESS` means the server has executed the statement while a status of `QUEUED` means the statement has been written to the command topic but the server has yet to execute it (and often means something is wrong). But you're right that these statuses aren't particularly actionable from the user's perspective, so we're probably better off nixing them from the client.